mpc-core/src/garble/label/state.rs

```diff
@@ -0,0 +1,137 @@
+use mpc_circuits::Value;
+
+use crate::garble::LabelError;
+
+use super::{Delta, WireLabel, WireLabelPair};
+
+mod sealed {
+    pub trait Sealed {}
+
+    impl Sealed for super::Full {}
+    impl Sealed for super::Active {}
+}
+
+/// Marker trait for label state
+pub trait State: sealed::Sealed {}
+
+/// Full wire labels
+#[derive(Debug, Clone, PartialEq)]
+pub struct Full {
+    /// Wire labels corresponding to logic low
+    pub(super) low: Vec<WireLabel>,
+    pub(super) delta: Delta,
+}
+
+impl State for Full {}
+
+impl Full {
+    pub(super) fn select(&self, value: &Value) -> Result<Active, LabelError> {
+        if value.len() != self.low.len() {
+            return Err(LabelError::InvalidValue(self.low.len(), value.len()));
+        }
+        Ok(Active {
+            labels: self
+                .low
+                .clone()
+                .into_iter()
+                .zip(value.to_lsb0_bits().into_iter())
+                .map(|(low, level)| if level { low ^ self.delta } else { low })
+                .collect(),
+        })
+    }
+
+    pub(super) fn to_labels(&self) -> Vec<WireLabelPair> {
+        self.low
+            .iter()
+            .map(|low| low.to_pair(self.delta, false))
+            .collect()
+    }
+
+    pub(super) fn from_decoding(
+        active: Active,
+        delta: Delta,
+        decoding: Vec<bool>,
```

Comment:
why is `decoding` not `Vec<LabelsDecodingInfo>`? It would be much cleaner, imo
It is only incidental that the decoding is a simple XOR operation. What if it was a complex function? We wouldn't want to copy-paste this function around.



This will require modifying Labels::decode() into Labels::decode_to_bits() and Labels::decode_to_group_value